spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamDefinition.java

```diff
@@ -127,25 +127,21 @@ public String toString() {
 
 
 	/**
-	 * Iterator that prevents mutation of its backing data structure.
-	 *
-	 * @param <T> the type of elements returned by this iterator
+	 * Iterator that walks through ModuleDefinitions in deployment order, that is, from downstream (sink, if present)
+	 * to upstream (source) module.
+	 * Also prevents mutation of its backing data structure.
 	 */
-	private static class ReadOnlyIterator<T> implements Iterator<T> {
-		private final Iterator<T> wrapped;
-
-		public ReadOnlyIterator(Iterator<T> wrapped) {
-			this.wrapped = wrapped;
-		}
+	public class DeploymentOrderIterator implements Iterator<ModuleDefinition> {
```

Comment:
It was made public to give access to its additional methods (like `hasMoreDownstream()`).
Since we've switched from enhanced for loop to while loop, I'd even make it **not** an iterator.

Actually, I feel like we're running in circles here, trying to shoehorn pretty complex logic (am I first?, am I last? are there named channels?) in the iterator pattern. And having vocabulary issues (the order being reversed, words like "first" and "last" become ambiguous) does not help.